internal/manifest: Add L0SubLevels methods to pick compactions

itsbilal opened the pull request:
This change adds methods to L0SubLevels to help pick, score,
and generate L0 -> LBase, and L0 -> L0 compactions, based on
information captured in the data structure about L0 sublevels.

Also adds associated datadriven unit tests, and a benchmark.

Covers a large part of #563. Thanks Sumeer for his work, most of
this was written by him.

First commit is #614 , which this PR builds on top of.

petermattis commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pebble/615)
<!-- Reviewable:end -->

itsbilal commented:
This PR is WIP until #614 is merged.

itsbilal commented:
This is ready for a review now. It's still meatier than I'd like (just under 1000 lines of non-test code changes), but I've added comments and/or simplified whatever I thought was worth simplifying. Also added tests for all the methods.

Happy to explain anything if necessary. It might be worth reading the test file first, and all the comments there, plus the giant comment before `PickBaseCompaction` to get the general idea.
